### ISSUE TITLE
capturing producer/consumer performance metrics

### ISSUE THREAD

**Issue by shakti-garg** [dpkp/kafka-python#2037]
I have written couple of producers/consumers but i want to capture their performance metrics in prometheus(or similar monitoring tool) for the sake of monitoring.
I have earlier done this using JMX on java-api for kafka but needs your guidance for this. I do see metrics.py in the code but not sure how to use this.

**Comment by tdentafix** [comment]
Hello,

Same issue here. We get a few hints of what's going on using the callbacks on `KafkaProducer` `send()` method but having metrics would be really better when handling high message rates. Indeed those require a lot of tuning to avoid any message losses and metrics can help.

From the documentation I understand that we have to implement the `AbstractMetricsReporter` class for which a squeleton is provided here : https://github.com/dpkp/kafka-python/blob/master/kafka/metrics/metrics_reporter.py

There are some comments in the methods which give a base idea of what we need to do but if any contributor could give a few more details about the life cycle and the arguments structure of the methods, that would really be appreciated.

Thanks a lot.

**Comment by jeffwidman** [comment]
https://kafka-python.readthedocs.io/en/master/apidoc/KafkaProducer.html#kafka.KafkaProducer.metrics (and similar one on the consumer side) are two functions that exist that you can call that will return metrics. 

The missing piece is exposing them via prometheus... for that you'll need a server exposing an endpoint that prometheus can scrape. You can either do that external to this lib--ie, call metrics periodically and report it via your app, or it does appear you can implement a  custom Metrics reporter class that handles the necessary plumbing to pipe the metrics where you want. In general, it'll be cleaner to write a custom reporter class.

I didn't write that section of the lib, and have never played with it, so can't help beyond that. If you do get it working I'd be curious to hear how it went for you.

**Comment by shakti-garg** [comment]
@jeffwidman I do end up exposing these metrics to prometheus using first approach. The output json from KafkaProducer.metrics() / KafkaConsumer.metrics() is mapped into prometheus-metrics format(wrote an utility: https://github.com/shakti-garg/prometheus_kafka_metrics.git) and then exposed at HTTP endpoint for prometheus to scrape. For now, i have done it for some services running on confluent-kafka-python library.

I agree second approach is more clean and hopefully, more efficient. I will give it a try.

**Comment by tdentafix** [comment]
Hello again,

Thanks a lot @jeffwidman and @shakti-garg for your insights.

I tried to play with the `AbstractMetricsReporter` class, but I don't see how I am supposed to get regular values of the metrics.

Indeed I do get all the `KafkaMetrics` objects at their creation when `metric_change()` is called, but after that, this method is never called again. I apparently have to call the `value()` method of the `KafkaMetric` objects I receive to get an updated value ...

I saw a "sensor" object in the `Metric()` class that might be the answer but I thought it was internal.

Anyway thanks for any further explanation.
